apollo-router/src/plugins/traffic_shaping/retry.rs

```diff
@@ -30,21 +29,57 @@ impl RetryPolicy {
                 retry_percent.unwrap_or(0.2),
             )),
             retry_mutations: retry_mutations.unwrap_or(false),
-            subgraph_name,
         }
     }
 }
 
-impl<Res, E> Policy<subgraph::Request, Res, E> for RetryPolicy {
+impl<E> Policy<subgraph::Request, subgraph::Response, E> for RetryPolicy {
     type Future = future::Ready<Self>;
 
-    fn retry(&self, req: &subgraph::Request, result: Result<&Res, &E>) -> Option<Self::Future> {
+    fn retry(
+        &self,
+        req: &subgraph::Request,
+        result: Result<&subgraph::Response, &E>,
+    ) -> Option<Self::Future> {
+        let subgraph_name = req.subgraph_name.clone().unwrap_or_default();
         match result {
-            Ok(_) => {
-                // Treat all `Response`s as success,
-                // so deposit budget and don't retry...
-                self.budget.deposit();
-                None
+            Ok(resp) => {
+                if resp.response.status() >= http::StatusCode::BAD_REQUEST {
```

Comment:
Is this correct? I mean, what if the subgraph does something "weird" like returning a redirect?

i.e.: should this be:
```
if resp.response.status() != http::StatusCode::OK {
```
?

Reply:
I recently saw users using 302 status code as a correct status code and they were making some stuffs in rhai about this. So I prefer to be conservative and use 400 status code as a threshold

Reply:
It seems weird to me. If we go with this, then it means that we are encoding "weird" behaviour into the router and that might not be what some customers want.

My preference is to do the expected thing, rather than produce something which fits into a "weird" behaviour of a customer. See what other people think, since I may really be in the minority here.

Reply:
>It seems weird to me. If we go with this, then it means that we are encoding "weird" behaviour into the router and that might not be what some customers want.

I agree, but I also agree with what @goto-bus-stop why would we retry on 3XX ?

Reply:
To be honest: I find the concept of the retry module a bit weird. Maybe the best thing to do is to define **exactly** which codes we will retry on and document that?

@goto-bus-stop Says definitely on a 429. I'd probably also want to retry a 503. Apart for those two, I probably wouldn't want to retry.

Reply:
I like starting with a small selection of codes.

I'm second-guessing 429 because we don't support `Retry-After` in this implementation and maybe that's something that clients should be doing, rather than us? If we just retry instantly then there is a good chance that we get a 429 again.

Reply:
https://www.rfc-editor.org/rfc/rfc9110.html#name-retry-after mentions 3xx is legit for retry.

Reply:
I don't think it says that. It says that the user agent should wait `Retry-After` seconds before following the redirect, not that it should retry the original request:
> When sent with any [3xx (Redirection)](https://www.rfc-editor.org/rfc/rfc9110.html#status.3xx) response, Retry-After indicates the minimum time that the user agent is asked to wait before issuing the redirected request.

Reply:
Would we want to do that? Retry to the redirected request?